glean-core/ios/Glean/Metrics/Ping.swift

```diff
@@ -82,6 +96,15 @@ public class Ping<ReasonCodesEnum: ReasonCodes>: PingBase {
     /// - parameters:
     ///     * reason: The reason the ping is being submitted.
     public func submit(reason: ReasonCodesEnum? = nil) {
+        if let cb = self.testCallback {
+            do {
+                try cb(reason)
+            } catch {
+                assert(false, "Callback threw before submitting ping \(name).")
```

Comment:
Huh. Well this is an interesting wrinkle in the semantics. Should we try to submit irrespective of thrown exceptions (like swift), or should we allow thrown exception to abort the submit (like Kotlin, Python)?

Either way we should be consistent. I'm leaning towards always submitting because I don't think granting an ability to suppress submission (even if solely in some nebulous "test mode") is a good idea.

Reply:
I added this merely to make it easier on users.
In Swift one needs to explicitly annotate functions as `throwable` if they do.
`testGetValue` can throw, so users need to use `try metric.testGetValue()` everywhere. That is fine in test methods, because they are throwable to begin with and will mark the test as failed.

Now when folks use `testBeforeNextSubmit` they should be able to use similar code to test their metrics. They are only allowed to call `testBeforeNextSubmit` in test mode (as enforced by the `assertInTestingMode()` call).

If we say that the closure a user passes is not throwable, than users need to wrap their `try metric.testGetValue()` in a `do {} catch {}` block themselves, only to then also fail the test.
Therefore I think the better ergonomics are if we do the wrapping and assert to fail tests if that happens.

Reply:
Ah, so we have equivalent ergo without having equivalent semantics. Fair enough.